### PR TITLE
Raise on File.stream's aenter if the file is a dir or doesn't exist

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
This matches the behaviour seen for URL resources (which happen to be the only other kind where the resource may not exist or may be invalid) and should solve the issue of aiohttp giving off confusing errors on unknown files  by erroring before we enter an aiohttp context.

### Checklist

- [ ] I have run `nox` and all the pipelines have passed.
- [ ] I have made unittests according to the code I have added/modified/deleted.

### Related issues
Fixes: #826

### CODE REVIEW

**Review by davfsa — COMMENTED** [review]
I am not overly happy with this. It does provide a more helpful error, but if people now how to debug python, they should be able to notice the file not found error above the error to send the request.

Additionally, this doesnt guarantee that the file is not deleted or moved after it is opened, which can lead to even more errors.

I would rather just have `aiohttp` deal with this and not us.

File: hikari/files.py [hunk 1]
```diff
@@ -768,6 +771,54 @@ class FileReader(AsyncReader, abc.ABC):
     """The path to the resource to read."""
 
 
+def _stat(path: pathlib.Path) -> os.stat_result:
+    # While paths will be implicitly resolved, we still need to explicitly
+    # call expanduser to deal with a ~ base.
+    try:
+        path = path.expanduser()
+    except RuntimeError:
+        pass  # A home directory couldn't be resolved, so we'll just use the path as-is.
```
[the remark this answers is not in the document]
Comment (FasterSpeeding) [reply hikari/files.py:780]: yeah no resolve doesn't expanduser on all platforms, that's Posix specific behaviour and on windows it'll just assume ~/ paths are relative to the cwd

**Comment by FasterSpeeding** [comment]
> I am not overly happy with this. It does provide a more helpful error, but if people now how to debug python, they should be able to notice the file not found error above the error to send the request.

Ok but this just matches the behaviour for URL where if it doesn't exist it'll error before being read into, giving a guarantee t that `file.__aenter__` will error if a resource doesn't exist instead of making that dependent on whether its a URL or File

> Additionally, this doesnt guarantee that the file is not deleted or moved after it is opened, which can lead to even more errors.

It's OS impl detail on whether that even matters anyways: unix systems will keep the file alive until the last application closes it and on windows you have to force delete it which at that point you're doing it to yourself; what we can do to handle this is keep the file pointer alive instead of re-opening it once they start reading so all that'd really need to be changed is to keep it open until its finished reading it rather than re-opening it which should be an easy change so yeah I'll do that lol

**Comment by FasterSpeeding** [comment]
I actually have an impl which keeps the file open for thy duration of the context manager for thread based file readers locally already but I don't think the same behaviour can be achieved with a process file reader and I'm also not sure what the point of allowing a process pool be used for file reading is (like file reading is blocking IO, not cpu bound work)

File: hikari/files.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (davfsa) [reply hikari/files.py:780]: Interesting. Didnt know this was a thing

**Comment by davfsa** [comment]
If you think it is worth having, then we can go for it

File: hikari/files.py [hunk 2]
```diff
@@ -768,6 +771,54 @@ class FileReader(AsyncReader, abc.ABC):
...
+        pass
+
+
+def _open_file(path: pathlib.Path) -> typing.BinaryIO:
+    # While paths will be implicitly resolved, we still need to explicitly
+    # call expanduser to deal with a ~ base.
+    try:
+        path = path.expanduser()
+    except RuntimeError:
+        pass  # A home directory couldn't be resolved, so we'll just use the path as-is.
```
[the remark this answers is not in the document]
Comment (FasterSpeeding) [reply hikari/files.py:817]: image

And idk if expanduser does anything blocking or not on any OS; I just kept calls to it within executor calls like the previous code was doing but added a try, except to account for the documented behaviour of "If a home directory can’t be resolved, [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError) is raised." and made sure that it's being consistently called for both ThreadedFileReader and MultiprocessingFileReader

[the remark this answers is not in the document]
Comment (davfsa) [reply hikari/files.py:817]: Ok, then looks good to me.

**Review by davfsa — APPROVED** [review]

**Pull request merged**